rust/kernel/tracepoint.rs

```diff
@@ -0,0 +1,80 @@
+// SPDX-License-Identifier: GPL-2.0
+
+// Copyright (C) 2024 Google LLC.
+
+//! Logic for tracepoints.
+
+/// Declare the Rust entry point for a tracepoint.
+#[macro_export]
+macro_rules! declare_trace {
+    ($(#[$attr:meta])* $pub:vis fn $name:ident($($argname:ident : $argtyp:ty),* $(,)?);) => {
+        $crate::macros::paste! {
+            $( #[$attr] )*
+            #[inline(always)]
+            $pub fn [< trace_ $name >]($($argname : $argtyp),*) {
+                #[cfg(CONFIG_TRACEPOINTS)]
+                {
+                    use $crate::bindings::*;
+
+                    let should_trace = unsafe {
+                        $crate::static_key::static_key_false!(
+                            [< __tracepoint_ $name >],
+                            $crate::bindings::tracepoint,
+                            key
+                        )
+                    };
+
+                    if should_trace {
+                        // TODO: cpu_online(raw_smp_processor_id())
```

Comment:
What are the ramifications of not having this? If the CPU is offline, will we hang, or just delay?